modules/monitoring/models/scheduledate.php

```diff
@@ -1,479 +1,485 @@
-		'servicegroups' => 'SCHEDULE_SERVICEGROUP_SVC_DOWNTIME'
-); # will schedule downtime for all services - not their hosts!
-	return $downtime_commands[$report_type];
-}
 
-/**
-* Save/update a recurring schedule
-*
-* @param $data array
-* @param $id int
-* @throws Exception
-*/
-public function edit_schedule($data, &$id = false) {
-	if (!$data) {
-		throw new Exception("Missing data for editing a scheduled downtime");
+	/**
+	 * Returns appropriate nagios command
+	 *
+	 * @param $type string
+	 * @return string
+	 */
+	static protected function get_nagios_cmd($type)
+	{
+		if (empty($type)) {
+			return false;
+		}
+		$downtime_commands = array(
+			'hosts' => 'SCHEDULE_HOST_DOWNTIME',
+			'services' => 'SCHEDULE_SVC_DOWNTIME',
+			'hostgroups' => 'SCHEDULE_HOSTGROUP_HOST_DOWNTIME',
+			'servicegroups' => 'SCHEDULE_SERVICEGROUP_SVC_DOWNTIME'
+		); # will schedule downtime for all services - not their hosts!
+		return $downtime_commands[$type];
 	}
-	foreach (static::$valid_fields as $field) {
-		if (!isset($data[$field])) {
-			throw new Exception("Missing field $field");
+
+	/**
+	 * Schedule a downtime by submitting it to nagios
+	 *
+	 * @param $objects array
+	 * @param $object_type string
+	 * @param $start_time string
+	 * @param $end_time string
+	 * @param $start_date string
+	 * @param $end_date string
+	 * @param $fixed string
+	 * @param $duration string
+	 * @param $comment string
+	 * @return boolean
+	 **/
+	public static function insert_downtimes($objects, $object_type, $start_time, $end_time, $start_date, $end_date, $fixed, $duration, $comment)
+	{
+		$result = array();
+		$nagios_cmd = self::determine_downtimetype($object_type);
+		$author = op5auth::instance()->get_user()->get_username();
+		$month = date('n');
+		$day = date('d');
+		$year = date('Y');
+		$strt_d = explode('-',$start_date);
+		$sy = (int)$strt_d[0];
+		$sm = (int)$strt_d[1];
+		$sd = (int)$strt_d[2];
+		$end_d = explode('-',$end_date);
+		$ey = (int)$end_d[0];
+		$em = (int)$end_d[1];
+		$ed = (int)$end_d[2];
+		$start_time = mktime(0, 0, self::time_to_seconds($start_time), $sm, $sd, $sy);
+		$end_time = mktime(0, 0, self::time_to_seconds($end_time), $em, $ed, $ey);
+		foreach ($objects as $object) {
+			if (static::check_if_scheduled($object_type, $object, $start_time, $end_time, $fixed)) {
+				// Skip object if it is already scheduled for downtime
+				continue;
+			}
+			$tmp_cmd = "$nagios_cmd;$object;$start_time;$end_time;$fixed;0;$duration;$author;AUTO: $comment";
+			$result[] = nagioscmd::submit_to_nagios($tmp_cmd);
 		}
+		return !in_array(false, $result);
 	}
-	$db = Database::instance();
-	$downtime_type = $data['downtime_type'];
-	if (!in_array($downtime_type, static::$valid_types, true)) {
-		throw new Exception("Downtime type $downtime_type is invalid");
+
+	/**
+	 * Given a time-like string (hh[:mm[:ss]]),
+	 * return the number of seconds involved.
+	 */
+	static public function time_to_seconds($time)
```

Comment:
Seems like php should have a way to do this, perhaps with [date()](https://www.php.net/manual/en/function.date.php) ?

Reply:
That code wasn't introduced with this commit - but yes, I agree.